{
 "cells": [
  {
   "cell_type": "raw",
   "id": "e3953639-82c0-4f3d-b34e-8219c1e0561f",
   "metadata": {},
   "source": [
    "import pandas as pd\n",
    "import csv\n",
    "import os\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3aa55df-1aba-4e3f-939e-afaa2c5a479d",
   "metadata": {},
   "source": [
    "# Load the data files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9a2e09a2-9a42-4b17-8105-d7d7559e3532",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGeneWithVer</th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>GeneDetail.refGeneWithVer</th>\n",
       "      <th>ExonicFunc.refGeneWithVer</th>\n",
       "      <th>AAChange.refGeneWithVer</th>\n",
       "      <th>...</th>\n",
       "      <th>D177F1_B1</th>\n",
       "      <th>D177M1_B1</th>\n",
       "      <th>D177P1_B1</th>\n",
       "      <th>D178F1_B1</th>\n",
       "      <th>D178M1_B1</th>\n",
       "      <th>D178P1_B1</th>\n",
       "      <th>D179F1_B1</th>\n",
       "      <th>D179M1_B1</th>\n",
       "      <th>D179P1_B1</th>\n",
       "      <th>Column_369</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924568</td>\n",
       "      <td>924568</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.137T&gt;A:p.L46H,S...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D125P1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924573</td>\n",
       "      <td>924573</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.142C&gt;T:p.P48S,S...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D132P2_B1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows Ã— 369 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Chr   Start     End Ref Alt Func.refGeneWithVer Gene.refGeneWithVer  \\\n",
       "0  chr1  924568  924568   T   A              exonic              SAMD11   \n",
       "1  chr1  924573  924573   C   T              exonic              SAMD11   \n",
       "\n",
       "  GeneDetail.refGeneWithVer ExonicFunc.refGeneWithVer  \\\n",
       "0                         .         nonsynonymous SNV   \n",
       "1                         .         nonsynonymous SNV   \n",
       "\n",
       "                             AAChange.refGeneWithVer  ...  \\\n",
       "0  \"SAMD11:NM_001385640.1:exon1:c.137T>A:p.L46H,S...  ...   \n",
       "1  \"SAMD11:NM_001385640.1:exon1:c.142C>T:p.P48S,S...  ...   \n",
       "\n",
       "             D177F1_B1            D177M1_B1            D177P1_B1  \\\n",
       "0  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"   \n",
       "1  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"   \n",
       "\n",
       "             D178F1_B1            D178M1_B1            D178P1_B1  \\\n",
       "0  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"   \n",
       "1  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"   \n",
       "\n",
       "             D179F1_B1            D179M1_B1            D179P1_B1 Column_369  \n",
       "0  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  D125P1_B1  \n",
       "1  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  \"./.:0,0:0:0:0,0,0\"  D132P2_B1  \n",
       "\n",
       "[2 rows x 369 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Myh11</td>\n",
       "      <td>6.746602e-21</td>\n",
       "      <td>-4.883850</td>\n",
       "      <td>0.217</td>\n",
       "      <td>0.973</td>\n",
       "      <td>1.014217e-16</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Actg2</td>\n",
       "      <td>1.188178e-20</td>\n",
       "      <td>-3.982674</td>\n",
       "      <td>0.326</td>\n",
       "      <td>0.991</td>\n",
       "      <td>1.786189e-16</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene         p_val  avg_log2FC  pct.1  pct.2     p_val_adj bladder_region\n",
       "0  Myh11  6.746602e-21   -4.883850  0.217  0.973  1.014217e-16          Whole\n",
       "1  Actg2  1.188178e-20   -3.982674  0.326  0.991  1.786189e-16          Whole"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>p_val_adj_fdr</th>\n",
       "      <th>cluster</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Myh11</td>\n",
       "      <td>2.712271e-08</td>\n",
       "      <td>-5.201319</td>\n",
       "      <td>0.154</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>1</td>\n",
       "      <td>Outer</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mylk</td>\n",
       "      <td>2.907453e-08</td>\n",
       "      <td>-2.759718</td>\n",
       "      <td>0.692</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>0.000051</td>\n",
       "      <td>1</td>\n",
       "      <td>Outer</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene         p_val  avg_log2FC  pct.1  pct.2  p_val_adj  p_val_adj_fdr  \\\n",
       "0  Myh11  2.712271e-08   -5.201319  0.154    1.0   0.000408       0.000051   \n",
       "1   Mylk  2.907453e-08   -2.759718  0.692    1.0   0.000437       0.000051   \n",
       "\n",
       "   cluster bladder_region  \n",
       "0        1          Outer  \n",
       "1        1          Outer  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>p_val_adj_fdr</th>\n",
       "      <th>cluster</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Ibsp</td>\n",
       "      <td>4.150000e-06</td>\n",
       "      <td>-4.123217</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.571</td>\n",
       "      <td>0.062419</td>\n",
       "      <td>0.002589</td>\n",
       "      <td>0</td>\n",
       "      <td>Intermediate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Cnn1</td>\n",
       "      <td>1.140000e-07</td>\n",
       "      <td>-3.427077</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.762</td>\n",
       "      <td>0.001716</td>\n",
       "      <td>0.000160</td>\n",
       "      <td>0</td>\n",
       "      <td>Intermediate</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   gene         p_val  avg_log2FC  pct.1  pct.2  p_val_adj  p_val_adj_fdr  \\\n",
       "0  Ibsp  4.150000e-06   -4.123217  0.000  0.571   0.062419       0.002589   \n",
       "1  Cnn1  1.140000e-07   -3.427077  0.038  0.762   0.001716       0.000160   \n",
       "\n",
       "   cluster bladder_region  \n",
       "0        0   Intermediate  \n",
       "1        0   Intermediate  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>p_val_adj_fdr</th>\n",
       "      <th>cluster</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Col10a1</td>\n",
       "      <td>2.402534e-07</td>\n",
       "      <td>-5.35406</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.947</td>\n",
       "      <td>0.003612</td>\n",
       "      <td>0.002761</td>\n",
       "      <td>2</td>\n",
       "      <td>Vascular</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Mmp13</td>\n",
       "      <td>7.822386e-07</td>\n",
       "      <td>-4.27050</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.895</td>\n",
       "      <td>0.011759</td>\n",
       "      <td>0.004496</td>\n",
       "      <td>2</td>\n",
       "      <td>Vascular</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene         p_val  avg_log2FC  pct.1  pct.2  p_val_adj  p_val_adj_fdr  \\\n",
       "0  Col10a1  2.402534e-07    -5.35406    0.0  0.947   0.003612       0.002761   \n",
       "1    Mmp13  7.822386e-07    -4.27050    0.0  0.895   0.011759       0.004496   \n",
       "\n",
       "   cluster bladder_region  \n",
       "0        2       Vascular  \n",
       "1        2       Vascular  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>p_val</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>pct.1</th>\n",
       "      <th>pct.2</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>p_val_adj_fdr</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Col3a1</td>\n",
       "      <td>3.273982e-09</td>\n",
       "      <td>-0.804882</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000049</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Krt10</td>\n",
       "      <td>5.650120e-09</td>\n",
       "      <td>-1.121357</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>0.000032</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene         p_val  avg_log2FC  pct.1  pct.2  p_val_adj  p_val_adj_fdr  \\\n",
       "0  Col3a1  3.273982e-09   -0.804882    1.0    1.0   0.000049       0.000032   \n",
       "1   Krt10  5.650120e-09   -1.121357    1.0    1.0   0.000085       0.000032   \n",
       "\n",
       "  bladder_region  \n",
       "0         subSMC  \n",
       "1         subSMC  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "wes_raw = pd.read_csv(\"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/WES-qiushi/updated_annovar_processed.txt\",\n",
    "                       sep='\\t',quoting=csv.QUOTE_NONE)\n",
    "display(wes_raw.head(2))\n",
    "\n",
    "visium_dea_filelist = [\"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/res_0.8/Whole_bladder_MGBvsWT_Filtered.csv\",\n",
    "            \"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/res_0.8/Outer_layer_Filtered_diffgenes_fdr_1MGBvsWT.csv\",\n",
    "            \"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/res_0.8/Intermediate_Filtered_diffgenes_fdr_0MGBvsWT.csv\",\n",
    "            \"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/res_0.8/Vascular_Filtered_diffgenes_fdr_2MGBvsWT.csv\",\n",
    "            \"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/res_0.8/subSMC_Filtered_fdr_SMCvsOther.csv\"]\n",
    "\n",
    "def combine_visium_dea_data(filelist):\n",
    "    \"\"\" \n",
    "    Combines multiple Visium Differential Expression Analysis (DEA) CSV files into a single DataFrame.\n",
    "    \n",
    "    Reads multiple CSV files, assigns each file a spatial region based on its filename, \n",
    "    and merges them into a single DataFrame with an added 'bladder_region' column.\n",
    "\n",
    "    Parameters:\n",
    "        filelist (list): A list of file paths to Visium DEA data CSV files.\n",
    "    \n",
    "    Returns:\n",
    "        pd.DataFrame: A concatenated DataFrame containing all DEA data with an additional 'bladder_region' column.\n",
    "    \"\"\"\n",
    "    visium = {} \n",
    "    for file in filelist:\n",
    "        basename = os.path.basename(file)\n",
    "        spatial_region = basename.split(\"_\")[0]\n",
    "        visium[spatial_region] = pd.read_csv(file)\n",
    "        visium[spatial_region][\"bladder_region\"] = spatial_region\n",
    "        display(visium[spatial_region].head(2))\n",
    "    combined = pd.concat(visium.values())\n",
    "    return(combined)\n",
    "\n",
    "def grab_visium_dea(gene, data):\n",
    "    \"\"\"\n",
    "    Extracts differential expression data for a specific gene from a Visium DEA DataFrame.\n",
    "\n",
    "    Filters the provided data to return only the rows corresponding to the given gene, \n",
    "    while omitting certain statistical columns.\n",
    "\n",
    "    Parameters:\n",
    "        gene (str): The gene of interest.\n",
    "        data (pd.DataFrame): A DataFrame containing Visium DEA data.\n",
    "    \n",
    "    Returns:\n",
    "        pd.DataFrame: A DataFrame with DEA results for the specified gene, excluding selected columns.\n",
    "    \"\"\"\n",
    "    df = pd.DataFrame()\n",
    "    if \"gene\" not in data.columns:\n",
    "        data.reset_index().rename(columns={df.index.name:'gene'})\n",
    "    cols_2_omit = [\"p_val\", \"pct.1\", \"pct.2\"]\n",
    "    sliced = data[data[\"gene\"] == gene]\n",
    "    mask = ~sliced.columns.isin(cols_2_omit)\n",
    "    sliced = sliced.loc[:, mask]\n",
    "    return(sliced)\n",
    "\n",
    "visium_data = combine_visium_dea_data(visium_dea_filelist)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff2cef8f-0ed1-4d00-ad34-47aff117b52f",
   "metadata": {},
   "source": [
    "## Exrtract only useful data to combine with WES"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2325881f-4430-4af8-a720-c6c642ad1bd8",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/yy/kt_303sd62zdch7grj7q09y00000gq/T/ipykernel_37752/820674496.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sliced['gene'] = sliced['gene'].str.upper()\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MYH11</td>\n",
       "      <td>-4.883850</td>\n",
       "      <td>1.014217e-16</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ACTG2</td>\n",
       "      <td>-3.982674</td>\n",
       "      <td>1.786189e-16</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACTA2</td>\n",
       "      <td>-3.083006</td>\n",
       "      <td>1.287243e-14</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MYL9</td>\n",
       "      <td>-2.544347</td>\n",
       "      <td>5.061845e-13</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MYLK</td>\n",
       "      <td>-2.767663</td>\n",
       "      <td>5.469357e-12</td>\n",
       "      <td>Whole</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>PTP4A1</td>\n",
       "      <td>0.677543</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>TACSTD2</td>\n",
       "      <td>-1.742884</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>ITGA5</td>\n",
       "      <td>0.900331</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>PARVB</td>\n",
       "      <td>-2.154998</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>DMPK</td>\n",
       "      <td>1.016926</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>subSMC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>210 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  avg_log2FC     p_val_adj bladder_region\n",
       "0     MYH11   -4.883850  1.014217e-16          Whole\n",
       "1     ACTG2   -3.982674  1.786189e-16          Whole\n",
       "2     ACTA2   -3.083006  1.287243e-14          Whole\n",
       "3      MYL9   -2.544347  5.061845e-13          Whole\n",
       "4      MYLK   -2.767663  5.469357e-12          Whole\n",
       "..      ...         ...           ...            ...\n",
       "40   PTP4A1    0.677543  1.000000e+00         subSMC\n",
       "41  TACSTD2   -1.742884  1.000000e+00         subSMC\n",
       "42    ITGA5    0.900331  1.000000e+00         subSMC\n",
       "43    PARVB   -2.154998  1.000000e+00         subSMC\n",
       "44     DMPK    1.016926  1.000000e+00         subSMC\n",
       "\n",
       "[210 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_2_extract = [\"gene\", \"avg_log2FC\", \"p_val_adj\", \"bladder_region\"]\n",
    "mask = visium_data.columns.isin(cols_2_extract)\n",
    "sliced = visium_data.loc[:, mask]\n",
    "sliced['gene'] = sliced['gene'].str.upper()\n",
    "sliced"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9cdc4fb2-09d8-4d0e-98a9-44cfe6cd0623",
   "metadata": {},
   "source": [
    "# Get Patient counts for each DEA genes from WES data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "896d8a98-fe09-4541-8696-3e8316dafb1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import re\n",
    "\n",
    "wes_data = wes_raw.copy(deep=True)\n",
    "\n",
    "# Identify patient and family columns\n",
    "total_Dcols = len([col for col in wes_data.columns if re.match(r'^D[0-9]', col)])\n",
    "patient_cols = [col for col in wes_data.columns if re.match(r'^D\\d{3}P', col)]  # D followed by 3 digits and P (e.g., D123P)\n",
    "family_cols = [col for col in wes_data.columns if re.match(r'^D\\d{3}(?!P)', col)]       # D followed by 3 digits but not Pafter that (e.g., D001M*)\n",
    "\n",
    "assert total_Dcols == len(patient_cols) + len(family_cols), \"number of columns not totalling up correctly\"\n",
    "\n",
    "# Extract genotype prefixes (strip quotes and split at colon)\n",
    "patient_genotypes = wes_data[patient_cols].replace(\n",
    "    r'^\"?(.*?):.*',  # Strip quotes and capture text before first colon\n",
    "    r'\\1', \n",
    "    regex=True\n",
    ")\n",
    "family_genotypes = wes_data[family_cols].replace(\n",
    "    r'^\"?(.*?):.*', \n",
    "    r'\\1', \n",
    "    regex=True\n",
    ")\n",
    "\n",
    "# Define invalid genotypes\n",
    "invalid_genotypes = ['./.', '0/0', '0|0']\n",
    "\n",
    "# Create masks for valid genotypes\n",
    "p_valid_mask = patient_genotypes.notna() & ~patient_genotypes.isin(invalid_genotypes)\n",
    "f_valid_mask = family_genotypes.notna() & ~family_genotypes.isin(invalid_genotypes) \n",
    "\n",
    "# Function to generate comma-separated IDs\n",
    "def get_ids(row_mask):\n",
    "    return ', '.join(row_mask.index[row_mask])\n",
    "\n",
    "wes_data['patient_count'] = p_valid_mask.sum(axis=1)\n",
    "wes_data['patient_ids'] = p_valid_mask.apply(get_ids, axis=1)\n",
    "wes_data['family_count'] = f_valid_mask.sum(axis=1)\n",
    "wes_data['family_ids'] = f_valid_mask.apply(get_ids, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4423956-52ee-41a7-aa16-f0831c47a7ba",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Ref</th>\n",
       "      <th>Alt</th>\n",
       "      <th>Func.refGeneWithVer</th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>GeneDetail.refGeneWithVer</th>\n",
       "      <th>ExonicFunc.refGeneWithVer</th>\n",
       "      <th>AAChange.refGeneWithVer</th>\n",
       "      <th>...</th>\n",
       "      <th>D178M1_B1</th>\n",
       "      <th>D178P1_B1</th>\n",
       "      <th>D179F1_B1</th>\n",
       "      <th>D179M1_B1</th>\n",
       "      <th>D179P1_B1</th>\n",
       "      <th>Column_369</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_count</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924568</td>\n",
       "      <td>924568</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.137T&gt;A:p.L46H,S...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D125P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D125P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924573</td>\n",
       "      <td>924573</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.142C&gt;T:p.P48S,S...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D132P2_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D132P2_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924628</td>\n",
       "      <td>924628</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.197C&gt;T:p.P66L,S...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D132P2_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D132P2_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924733</td>\n",
       "      <td>924733</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.302C&gt;T:p.A101V,...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D078F1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "      <td>1</td>\n",
       "      <td>D078F1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>924930</td>\n",
       "      <td>924930</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>exonic</td>\n",
       "      <td>SAMD11</td>\n",
       "      <td>.</td>\n",
       "      <td>nonsynonymous SNV</td>\n",
       "      <td>\"SAMD11:NM_001385640.1:exon1:c.499C&gt;A:p.R167S,...</td>\n",
       "      <td>...</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>\"./.:0,0:0:0:0,0,0\"</td>\n",
       "      <td>D133P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D133P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33588</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154357624</td>\n",
       "      <td>154357624</td>\n",
       "      <td>-</td>\n",
       "      <td>GTGATCTGGACAGCCAGCAGGCCCTCCCCGGCGTCCTTTGCA</td>\n",
       "      <td>splicing</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>NM_001110556.2:exon29:c.4756-1-&gt;TGCAAAGGACGCCG...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>\"0/0:31,0:31:90:0,90,1113\"</td>\n",
       "      <td>\"0/0:31,0:31:90:0,90,1085\"</td>\n",
       "      <td>\"0/0:34,0:34:93:0,93,1161\"</td>\n",
       "      <td>\"0/0:32,0:32:90:0,90,1239\"</td>\n",
       "      <td>\"0/0:30,0:30:90:0,90,1234\"</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33589</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154358356</td>\n",
       "      <td>154358444</td>\n",
       "      <td>CTGTGAGGGATTGGTGTTGTGAGCAGTCAGACAGGTTCTCAGCATC...</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>NM_001110556.2:exon27:r.spl;NM_001110556.2:exo...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>\"0/0:31,0:31:90:.:.:0,90,1139:.\"</td>\n",
       "      <td>\"0/0:32,0:32:90:.:.:0,90,1167:.\"</td>\n",
       "      <td>\"0/0:30,0:30:90:.:.:0,90,1171:.\"</td>\n",
       "      <td>\"0/0:40,0:40:90:.:.:0,90,1320:.\"</td>\n",
       "      <td>\"0/0:34,0:34:90:.:.:0,90,1350:.\"</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33590</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154365260</td>\n",
       "      <td>154365348</td>\n",
       "      <td>CTGCCAAGACAAGGAGGGCCTCAGGCCTGCCCAGCAGTGAACCCGG...</td>\n",
       "      <td>-</td>\n",
       "      <td>splicing</td>\n",
       "      <td>FLNA</td>\n",
       "      <td>NM_001110556.2:exon10:r.spl;NM_001110556.2:exo...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>\"0/0:43,0:43:90:0,90,1350\"</td>\n",
       "      <td>\"0/0:33,0:33:90:0,90,1350\"</td>\n",
       "      <td>\"0/0:31,0:31:90:0,90,1063\"</td>\n",
       "      <td>\"0/0:31,0:31:90:0,90,1350\"</td>\n",
       "      <td>\"0/0:35,0:35:90:0,90,1289\"</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33591</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154405656</td>\n",
       "      <td>154405656</td>\n",
       "      <td>-</td>\n",
       "      <td>GTAG</td>\n",
       "      <td>splicing</td>\n",
       "      <td>DNASE1L1</td>\n",
       "      <td>NM_006730.4:exon2:UTR5;NM_001009934.2:exon2:UT...</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>\"0/0:30,0:30:90:.:.:0,90,1169:.\"</td>\n",
       "      <td>\"0/0:34,0:34:90:.:.:0,90,1236:.\"</td>\n",
       "      <td>\"0/0:30,0:30:90:.:.:0,90,1111:.\"</td>\n",
       "      <td>\"0/0:32,0:32:93:.:.:0,93,996:.\"</td>\n",
       "      <td>\"0/0:34,0:34:90:.:.:0,90,1350:.\"</td>\n",
       "      <td>D105P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D105P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33592</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154478059</td>\n",
       "      <td>154478059</td>\n",
       "      <td>-</td>\n",
       "      <td>GGT</td>\n",
       "      <td>splicing</td>\n",
       "      <td>LAGE3</td>\n",
       "      <td>NM_006014.4:exon3:c.318-1-&gt;ACC</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>...</td>\n",
       "      <td>\"0/0:35,0:35:96:.:.:0,96,1440:.\"</td>\n",
       "      <td>\"0/0:35,0:35:96:.:.:0,96,1440:.\"</td>\n",
       "      <td>\"0/0:31,0:31:93:.:.:0,93,1208:.\"</td>\n",
       "      <td>\"0/0:33,0:33:90:.:.:0,90,1350:.\"</td>\n",
       "      <td>\"0/0:32,0:32:90:.:.:0,90,1254:.\"</td>\n",
       "      <td>D099P1_B1</td>\n",
       "      <td>1</td>\n",
       "      <td>D099P1_B1</td>\n",
       "      <td>0</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33593 rows Ã— 373 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Chr      Start        End  \\\n",
       "0      chr1     924568     924568   \n",
       "1      chr1     924573     924573   \n",
       "2      chr1     924628     924628   \n",
       "3      chr1     924733     924733   \n",
       "4      chr1     924930     924930   \n",
       "...     ...        ...        ...   \n",
       "33588  chrX  154357624  154357624   \n",
       "33589  chrX  154358356  154358444   \n",
       "33590  chrX  154365260  154365348   \n",
       "33591  chrX  154405656  154405656   \n",
       "33592  chrX  154478059  154478059   \n",
       "\n",
       "                                                     Ref  \\\n",
       "0                                                      T   \n",
       "1                                                      C   \n",
       "2                                                      C   \n",
       "3                                                      C   \n",
       "4                                                      C   \n",
       "...                                                  ...   \n",
       "33588                                                  -   \n",
       "33589  CTGTGAGGGATTGGTGTTGTGAGCAGTCAGACAGGTTCTCAGCATC...   \n",
       "33590  CTGCCAAGACAAGGAGGGCCTCAGGCCTGCCCAGCAGTGAACCCGG...   \n",
       "33591                                                  -   \n",
       "33592                                                  -   \n",
       "\n",
       "                                              Alt Func.refGeneWithVer  \\\n",
       "0                                               A              exonic   \n",
       "1                                               T              exonic   \n",
       "2                                               T              exonic   \n",
       "3                                               T              exonic   \n",
       "4                                               A              exonic   \n",
       "...                                           ...                 ...   \n",
       "33588  GTGATCTGGACAGCCAGCAGGCCCTCCCCGGCGTCCTTTGCA            splicing   \n",
       "33589                                           -            splicing   \n",
       "33590                                           -            splicing   \n",
       "33591                                        GTAG            splicing   \n",
       "33592                                         GGT            splicing   \n",
       "\n",
       "      Gene.refGeneWithVer                          GeneDetail.refGeneWithVer  \\\n",
       "0                  SAMD11                                                  .   \n",
       "1                  SAMD11                                                  .   \n",
       "2                  SAMD11                                                  .   \n",
       "3                  SAMD11                                                  .   \n",
       "4                  SAMD11                                                  .   \n",
       "...                   ...                                                ...   \n",
       "33588                FLNA  NM_001110556.2:exon29:c.4756-1->TGCAAAGGACGCCG...   \n",
       "33589                FLNA  NM_001110556.2:exon27:r.spl;NM_001110556.2:exo...   \n",
       "33590                FLNA  NM_001110556.2:exon10:r.spl;NM_001110556.2:exo...   \n",
       "33591            DNASE1L1  NM_006730.4:exon2:UTR5;NM_001009934.2:exon2:UT...   \n",
       "33592               LAGE3                     NM_006014.4:exon3:c.318-1->ACC   \n",
       "\n",
       "      ExonicFunc.refGeneWithVer  \\\n",
       "0             nonsynonymous SNV   \n",
       "1             nonsynonymous SNV   \n",
       "2             nonsynonymous SNV   \n",
       "3             nonsynonymous SNV   \n",
       "4             nonsynonymous SNV   \n",
       "...                         ...   \n",
       "33588                         .   \n",
       "33589                         .   \n",
       "33590                         .   \n",
       "33591                         .   \n",
       "33592                         .   \n",
       "\n",
       "                                 AAChange.refGeneWithVer  ...  \\\n",
       "0      \"SAMD11:NM_001385640.1:exon1:c.137T>A:p.L46H,S...  ...   \n",
       "1      \"SAMD11:NM_001385640.1:exon1:c.142C>T:p.P48S,S...  ...   \n",
       "2      \"SAMD11:NM_001385640.1:exon1:c.197C>T:p.P66L,S...  ...   \n",
       "3      \"SAMD11:NM_001385640.1:exon1:c.302C>T:p.A101V,...  ...   \n",
       "4      \"SAMD11:NM_001385640.1:exon1:c.499C>A:p.R167S,...  ...   \n",
       "...                                                  ...  ...   \n",
       "33588                                                  .  ...   \n",
       "33589                                                  .  ...   \n",
       "33590                                                  .  ...   \n",
       "33591                                                  .  ...   \n",
       "33592                                                  .  ...   \n",
       "\n",
       "                              D178M1_B1                         D178P1_B1  \\\n",
       "0                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "1                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "2                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "3                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "4                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "...                                 ...                               ...   \n",
       "33588        \"0/0:31,0:31:90:0,90,1113\"        \"0/0:31,0:31:90:0,90,1085\"   \n",
       "33589  \"0/0:31,0:31:90:.:.:0,90,1139:.\"  \"0/0:32,0:32:90:.:.:0,90,1167:.\"   \n",
       "33590        \"0/0:43,0:43:90:0,90,1350\"        \"0/0:33,0:33:90:0,90,1350\"   \n",
       "33591  \"0/0:30,0:30:90:.:.:0,90,1169:.\"  \"0/0:34,0:34:90:.:.:0,90,1236:.\"   \n",
       "33592  \"0/0:35,0:35:96:.:.:0,96,1440:.\"  \"0/0:35,0:35:96:.:.:0,96,1440:.\"   \n",
       "\n",
       "                              D179F1_B1                         D179M1_B1  \\\n",
       "0                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "1                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "2                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "3                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "4                   \"./.:0,0:0:0:0,0,0\"               \"./.:0,0:0:0:0,0,0\"   \n",
       "...                                 ...                               ...   \n",
       "33588        \"0/0:34,0:34:93:0,93,1161\"        \"0/0:32,0:32:90:0,90,1239\"   \n",
       "33589  \"0/0:30,0:30:90:.:.:0,90,1171:.\"  \"0/0:40,0:40:90:.:.:0,90,1320:.\"   \n",
       "33590        \"0/0:31,0:31:90:0,90,1063\"        \"0/0:31,0:31:90:0,90,1350\"   \n",
       "33591  \"0/0:30,0:30:90:.:.:0,90,1111:.\"   \"0/0:32,0:32:93:.:.:0,93,996:.\"   \n",
       "33592  \"0/0:31,0:31:93:.:.:0,93,1208:.\"  \"0/0:33,0:33:90:.:.:0,90,1350:.\"   \n",
       "\n",
       "                              D179P1_B1 Column_369 patient_count patient_ids  \\\n",
       "0                   \"./.:0,0:0:0:0,0,0\"  D125P1_B1             1   D125P1_B1   \n",
       "1                   \"./.:0,0:0:0:0,0,0\"  D132P2_B1             1   D132P2_B1   \n",
       "2                   \"./.:0,0:0:0:0,0,0\"  D132P2_B1             1   D132P2_B1   \n",
       "3                   \"./.:0,0:0:0:0,0,0\"  D078F1_B1             0               \n",
       "4                   \"./.:0,0:0:0:0,0,0\"  D133P1_B1             1   D133P1_B1   \n",
       "...                                 ...        ...           ...         ...   \n",
       "33588        \"0/0:30,0:30:90:0,90,1234\"  D081P1_B1             1   D081P1_B1   \n",
       "33589  \"0/0:34,0:34:90:.:.:0,90,1350:.\"  D081P1_B1             1   D081P1_B1   \n",
       "33590        \"0/0:35,0:35:90:0,90,1289\"  D081P1_B1             1   D081P1_B1   \n",
       "33591  \"0/0:34,0:34:90:.:.:0,90,1350:.\"  D105P1_B1             1   D105P1_B1   \n",
       "33592  \"0/0:32,0:32:90:.:.:0,90,1254:.\"  D099P1_B1             1   D099P1_B1   \n",
       "\n",
       "      family_count family_ids  \n",
       "0                0             \n",
       "1                0             \n",
       "2                0             \n",
       "3                1  D078F1_B1  \n",
       "4                0             \n",
       "...            ...        ...  \n",
       "33588            0             \n",
       "33589            0             \n",
       "33590            0             \n",
       "33591            0             \n",
       "33592            0             \n",
       "\n",
       "[33593 rows x 373 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wes_data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "042f22ac-5696-45b4-b147-ae0edb5f9d85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene.refGeneWithVer</th>\n",
       "      <th>AlphaMissense_score</th>\n",
       "      <th>AlphaMissense_rankscore</th>\n",
       "      <th>AlphaMissense_pred</th>\n",
       "      <th>gnomad41_exome_AF</th>\n",
       "      <th>gnomad41_exome_faf99</th>\n",
       "      <th>gnomad41_exome_AF_raw</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>AlphaMissense_score</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>CADD_phred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0.299</td>\n",
       "      <td>0.52921</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.299</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>21.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.11893</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.088</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>22.30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.16843</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.099</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>18.18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.21011</td>\n",
       "      <td>B</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000011</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.72</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMD11</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.99395</td>\n",
       "      <td>P</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>33.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33588</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>0.000</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33589</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>0.000</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33590</th>\n",
       "      <td>FLNA</td>\n",
       "      <td>0.000</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33591</th>\n",
       "      <td>DNASE1L1</td>\n",
       "      <td>0.000</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33592</th>\n",
       "      <td>LAGE3</td>\n",
       "      <td>0.000</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>33593 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Gene.refGeneWithVer  AlphaMissense_score AlphaMissense_rankscore  \\\n",
       "0                  SAMD11                0.299                 0.52921   \n",
       "1                  SAMD11                0.088                 0.11893   \n",
       "2                  SAMD11                0.099                 0.16843   \n",
       "3                  SAMD11                0.110                 0.21011   \n",
       "4                  SAMD11                1.000                 0.99395   \n",
       "...                   ...                  ...                     ...   \n",
       "33588                FLNA                0.000                       .   \n",
       "33589                FLNA                0.000                       .   \n",
       "33590                FLNA                0.000                       .   \n",
       "33591            DNASE1L1                0.000                       .   \n",
       "33592               LAGE3                0.000                       .   \n",
       "\n",
       "      AlphaMissense_pred  gnomad41_exome_AF  gnomad41_exome_faf99  \\\n",
       "0                      B                0.0                   0.0   \n",
       "1                      B                0.0                   0.0   \n",
       "2                      B                0.0                   0.0   \n",
       "3                      B                0.0                   0.0   \n",
       "4                      P                0.0                   0.0   \n",
       "...                  ...                ...                   ...   \n",
       "33588                  .                0.0                   0.0   \n",
       "33589                  .                0.0                   0.0   \n",
       "33590                  .                0.0                   0.0   \n",
       "33591                  .                0.0                   0.0   \n",
       "33592                  .                0.0                   0.0   \n",
       "\n",
       "       gnomad41_exome_AF_raw  Polyphen2_HDIV_score  SIFT_score  \\\n",
       "0                   0.000029                   0.0         0.0   \n",
       "1                   0.000000                   0.0         0.0   \n",
       "2                   0.000006                   0.0         0.0   \n",
       "3                   0.000011                   0.0         0.0   \n",
       "4                   0.000013                   0.0         0.0   \n",
       "...                      ...                   ...         ...   \n",
       "33588               0.000000                   0.0         0.0   \n",
       "33589               0.000000                   0.0         0.0   \n",
       "33590               0.000000                   0.0         0.0   \n",
       "33591               0.000000                   0.0         0.0   \n",
       "33592               0.000000                   0.0         0.0   \n",
       "\n",
       "       AlphaMissense_score  MutationAssessor_score  MutationTaster_score  \\\n",
       "0                    0.299                     0.0                   0.0   \n",
       "1                    0.088                     0.0                   0.0   \n",
       "2                    0.099                     0.0                   0.0   \n",
       "3                    0.110                     0.0                   0.0   \n",
       "4                    1.000                     0.0                   0.0   \n",
       "...                    ...                     ...                   ...   \n",
       "33588                0.000                     0.0                   0.0   \n",
       "33589                0.000                     0.0                   0.0   \n",
       "33590                0.000                     0.0                   0.0   \n",
       "33591                0.000                     0.0                   0.0   \n",
       "33592                0.000                     0.0                   0.0   \n",
       "\n",
       "       CADD_phred  \n",
       "0           21.70  \n",
       "1           22.30  \n",
       "2           18.18  \n",
       "3           17.72  \n",
       "4           33.00  \n",
       "...           ...  \n",
       "33588        0.00  \n",
       "33589        0.00  \n",
       "33590        0.00  \n",
       "33591        0.00  \n",
       "33592        0.00  \n",
       "\n",
       "[33593 rows x 13 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols_dtype_convert = [\"gnomad41_exome_AF\", \"gnomad41_exome_faf99\", \"gnomad41_exome_AF_raw\", \"Polyphen2_HDIV_score\", \n",
    "                      \"SIFT_score\", \"AlphaMissense_score\", \"MutationAssessor_score\", \"MutationTaster_score\", \"CADD_phred\"]\n",
    "wes_data[[\"Gene.refGeneWithVer\"] + [i for i in wes_data.columns if re.match(r'Alpha', i)] + cols_dtype_convert]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbc9b8b5-930c-4397-a218-a474ebd7cb24",
   "metadata": {},
   "source": [
    "## Filter the WES data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e1df3469-12f5-4249-b4e5-e11461f324db",
   "metadata": {},
   "outputs": [],
   "source": [
    "cols_dtype_convert = [\"gnomad41_exome_AF\", \"gnomad41_exome_faf99\", \"gnomad41_exome_AF_raw\", \"Polyphen2_HDIV_score\", \"SIFT_score\", \n",
    "                      \"AlphaMissense_score\", \"MutationAssessor_score\", \"MutationTaster_score\", \"CADD_phred\"]\n",
    "wes_data[cols_dtype_convert] = wes_data[cols_dtype_convert].replace('.','0').astype(float)\n",
    "af_cut = 0.001# %0.1\n",
    "af_filtered = wes_data[wes_data[\"gnomad41_exome_AF\"] < af_cut] # AF < 0.001%\n",
    "\n",
    "# Conditions to test, we are including if values are zero because \n",
    "# 1. in the previous step we converted '.' tp 0 and \n",
    "# 2. The indels or frameshifts will have 0s in all the filters\n",
    "conditions = [\n",
    "    (af_filtered[\"Polyphen2_HDIV_score\"] > 0.9) | (af_filtered[\"Polyphen2_HDIV_score\"] == 0),          # Condition 1\n",
    "    (af_filtered[\"SIFT_score\"] < 0.1) | (af_filtered[\"SIFT_score\"] == 0),                              # Condition 2\n",
    "    (af_filtered[\"AlphaMissense_score\"] > 0.5) | (af_filtered[\"AlphaMissense_score\"] == 0),            # Condition 3\n",
    "    (af_filtered[\"MutationAssessor_score\"] > 1.90) | (af_filtered[\"MutationAssessor_score\"] == 0),     # Condition 4 # \n",
    "    (af_filtered[\"MutationTaster_score\"] > 0.5) | (af_filtered[\"MutationTaster_score\"] == 0),                                                         # Condition 5  \n",
    "    (af_filtered[\"CADD_phred\"] > 20) | (af_filtered[\"CADD_phred\"] == 0),                                                                    # Condition 6\n",
    "    \n",
    "]\n",
    "# Sum the True con ditions for each row (True = 1, False = 0)\n",
    "count_met = sum(conditions)\n",
    "# Create a mask where at least 4 conditions are True\n",
    "mask = count_met >= 5\n",
    "# Filter the DataFrame using the mask\n",
    "filtered_wes_data = af_filtered[mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "483fb2c8-2add-4e47-b1fc-07533b0e3903",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAChange.refGeneWithVer</th>\n",
       "      <th>gnomad41_exome_AF</th>\n",
       "      <th>gnomad41_exome_faf99</th>\n",
       "      <th>gnomad41_exome_AF_raw</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>AlphaMissense_score</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>CADD_phred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27311</th>\n",
       "      <td>\"CNN2:NM_001303499.2:exon2:c.104G&gt;A:p.R35H,CNN...</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.982</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.318</td>\n",
       "      <td>2.855</td>\n",
       "      <td>0.988787</td>\n",
       "      <td>23.7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27312</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.695C&gt;A:p.P232H,CNN2:...</td>\n",
       "      <td>0.000056</td>\n",
       "      <td>0.000042</td>\n",
       "      <td>0.004900</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.324</td>\n",
       "      <td>2.005</td>\n",
       "      <td>0.990314</td>\n",
       "      <td>26.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27314</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.720_721insAC:p.A241T...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27316</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.723_724del:p.D242Wfs...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 AAChange.refGeneWithVer  gnomad41_exome_AF  \\\n",
       "27311  \"CNN2:NM_001303499.2:exon2:c.104G>A:p.R35H,CNN...           0.000029   \n",
       "27312  \"CNN2:NM_201277.3:exon6:c.695C>A:p.P232H,CNN2:...           0.000056   \n",
       "27314  \"CNN2:NM_201277.3:exon6:c.720_721insAC:p.A241T...           0.000000   \n",
       "27316  \"CNN2:NM_201277.3:exon6:c.723_724del:p.D242Wfs...           0.000000   \n",
       "\n",
       "       gnomad41_exome_faf99  gnomad41_exome_AF_raw  Polyphen2_HDIV_score  \\\n",
       "27311              0.000020               0.000029                 0.982   \n",
       "27312              0.000042               0.004900                 0.998   \n",
       "27314              0.000000               0.000079                 0.000   \n",
       "27316              0.000000               0.000078                 0.000   \n",
       "\n",
       "       SIFT_score  AlphaMissense_score  MutationAssessor_score  \\\n",
       "27311       0.058                0.318                   2.855   \n",
       "27312       0.006                0.324                   2.005   \n",
       "27314       0.000                0.000                   0.000   \n",
       "27316       0.000                0.000                   0.000   \n",
       "\n",
       "       MutationTaster_score  CADD_phred  \n",
       "27311              0.988787        23.7  \n",
       "27312              0.990314        26.3  \n",
       "27314              0.000000         0.0  \n",
       "27316              0.000000         0.0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAChange.refGeneWithVer</th>\n",
       "      <th>gnomad41_exome_AF</th>\n",
       "      <th>gnomad41_exome_faf99</th>\n",
       "      <th>gnomad41_exome_AF_raw</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>AlphaMissense_score</th>\n",
       "      <th>MutationAssessor_score</th>\n",
       "      <th>MutationTaster_score</th>\n",
       "      <th>CADD_phred</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27311</th>\n",
       "      <td>\"CNN2:NM_001303499.2:exon2:c.104G&gt;A:p.R35H,CNN...</td>\n",
       "      <td>2.943000e-05</td>\n",
       "      <td>1.971000e-05</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.982</td>\n",
       "      <td>0.058</td>\n",
       "      <td>0.318</td>\n",
       "      <td>2.855</td>\n",
       "      <td>0.988787</td>\n",
       "      <td>23.700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27312</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.695C&gt;A:p.P232H,CNN2:...</td>\n",
       "      <td>5.560000e-05</td>\n",
       "      <td>4.213000e-05</td>\n",
       "      <td>0.004900</td>\n",
       "      <td>0.998</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.324</td>\n",
       "      <td>2.005</td>\n",
       "      <td>0.990314</td>\n",
       "      <td>26.300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27313</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.715A&gt;C:p.T239P,CNN2:...</td>\n",
       "      <td>6.911000e-07</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.002400</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.382</td>\n",
       "      <td>0.052</td>\n",
       "      <td>-0.205</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>5.225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27314</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.720_721insAC:p.A241T...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000079</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27315</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.721G&gt;A:p.A241T,CNN2:...</td>\n",
       "      <td>1.383000e-06</td>\n",
       "      <td>9.000000e-08</td>\n",
       "      <td>0.002300</td>\n",
       "      <td>0.501</td>\n",
       "      <td>0.524</td>\n",
       "      <td>0.068</td>\n",
       "      <td>1.590</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>17.750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27316</th>\n",
       "      <td>\"CNN2:NM_201277.3:exon6:c.723_724del:p.D242Wfs...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 AAChange.refGeneWithVer  gnomad41_exome_AF  \\\n",
       "27311  \"CNN2:NM_001303499.2:exon2:c.104G>A:p.R35H,CNN...       2.943000e-05   \n",
       "27312  \"CNN2:NM_201277.3:exon6:c.695C>A:p.P232H,CNN2:...       5.560000e-05   \n",
       "27313  \"CNN2:NM_201277.3:exon6:c.715A>C:p.T239P,CNN2:...       6.911000e-07   \n",
       "27314  \"CNN2:NM_201277.3:exon6:c.720_721insAC:p.A241T...       0.000000e+00   \n",
       "27315  \"CNN2:NM_201277.3:exon6:c.721G>A:p.A241T,CNN2:...       1.383000e-06   \n",
       "27316  \"CNN2:NM_201277.3:exon6:c.723_724del:p.D242Wfs...       0.000000e+00   \n",
       "\n",
       "       gnomad41_exome_faf99  gnomad41_exome_AF_raw  Polyphen2_HDIV_score  \\\n",
       "27311          1.971000e-05               0.000029                 0.982   \n",
       "27312          4.213000e-05               0.004900                 0.998   \n",
       "27313          0.000000e+00               0.002400                 0.001   \n",
       "27314          0.000000e+00               0.000079                 0.000   \n",
       "27315          9.000000e-08               0.002300                 0.501   \n",
       "27316          0.000000e+00               0.000078                 0.000   \n",
       "\n",
       "       SIFT_score  AlphaMissense_score  MutationAssessor_score  \\\n",
       "27311       0.058                0.318                   2.855   \n",
       "27312       0.006                0.324                   2.005   \n",
       "27313       0.382                0.052                  -0.205   \n",
       "27314       0.000                0.000                   0.000   \n",
       "27315       0.524                0.068                   1.590   \n",
       "27316       0.000                0.000                   0.000   \n",
       "\n",
       "       MutationTaster_score  CADD_phred  \n",
       "27311              0.988787      23.700  \n",
       "27312              0.990314      26.300  \n",
       "27313              1.000000       5.225  \n",
       "27314              0.000000       0.000  \n",
       "27315              1.000000      17.750  \n",
       "27316              0.000000       0.000  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "filtered_gene = filtered_wes_data[filtered_wes_data[\"Gene.refGeneWithVer\"] == \"CNN2\"]\n",
    "display(filtered_gene[[\"AAChange.refGeneWithVer\"] + cols_dtype_convert])\n",
    "\n",
    "unfiltered_gene = wes_data[wes_data[\"Gene.refGeneWithVer\"] == \"CNN2\"]\n",
    "display(unfiltered_gene[[\"AAChange.refGeneWithVer\"] + cols_dtype_convert])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8209b385-917a-44db-aef9-bd25f1e89230",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Intermediate\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "      <th>index</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>family_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MYH11</th>\n",
       "      <td>-3.397872</td>\n",
       "      <td>4.670000e-07</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D013P1_B1</td>\n",
       "      <td>D013M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTG2</th>\n",
       "      <td>-3.296180</td>\n",
       "      <td>8.950000e-07</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>D034P1_B1, D074P1_B1, D065P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>POLR1A</th>\n",
       "      <td>-2.193822</td>\n",
       "      <td>8.350236e-02</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>2502</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>D040F1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYLK</th>\n",
       "      <td>-2.028758</td>\n",
       "      <td>1.512670e-04</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D165P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL9</th>\n",
       "      <td>-1.932520</td>\n",
       "      <td>5.070000e-06</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>2008</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D156P1_B1</td>\n",
       "      <td>D040M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ARHGEF11</th>\n",
       "      <td>-1.658454</td>\n",
       "      <td>1.152607e-01</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>236</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D164P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>IGFBP2</th>\n",
       "      <td>-1.467422</td>\n",
       "      <td>3.615104e-02</td>\n",
       "      <td>Intermediate</td>\n",
       "      <td>1479</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td></td>\n",
       "      <td>D109M2_B1, D109M2_B1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          avg_log2FC     p_val_adj bladder_region  index  patient_count  \\\n",
       "gene                                                                      \n",
       "MYH11      -3.397872  4.670000e-07   Intermediate   1997              1   \n",
       "ACTG2      -3.296180  8.950000e-07   Intermediate     47              3   \n",
       "POLR1A     -2.193822  8.350236e-02   Intermediate   2502              0   \n",
       "MYLK       -2.028758  1.512670e-04   Intermediate   2009              1   \n",
       "MYL9       -1.932520  5.070000e-06   Intermediate   2008              1   \n",
       "ARHGEF11   -1.658454  1.152607e-01   Intermediate    236              1   \n",
       "IGFBP2     -1.467422  3.615104e-02   Intermediate   1479              0   \n",
       "\n",
       "          family_count                      patient_ids            family_ids  \n",
       "gene                                                                           \n",
       "MYH11                1                        D013P1_B1             D013M1_B1  \n",
       "ACTG2                0  D034P1_B1, D074P1_B1, D065P1_B1                        \n",
       "POLR1A               1                                              D040F1_B1  \n",
       "MYLK                 0                        D165P1_B1                        \n",
       "MYL9                 1                        D156P1_B1             D040M1_B1  \n",
       "ARHGEF11             0                        D164P1_B1                        \n",
       "IGFBP2               2                                   D109M2_B1, D109M2_B1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(7, 8)\n",
      "Outer\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "      <th>index</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>family_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MYH11</th>\n",
       "      <td>-5.201319</td>\n",
       "      <td>0.000408</td>\n",
       "      <td>Outer</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D013P1_B1</td>\n",
       "      <td>D013M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYLK</th>\n",
       "      <td>-2.759718</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>Outer</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D165P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A2</th>\n",
       "      <td>1.253755</td>\n",
       "      <td>0.000439</td>\n",
       "      <td>Outer</td>\n",
       "      <td>687</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D150P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTG2</th>\n",
       "      <td>-3.668688</td>\n",
       "      <td>0.000524</td>\n",
       "      <td>Outer</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>D034P1_B1, D074P1_B1, D065P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL9</th>\n",
       "      <td>-2.319767</td>\n",
       "      <td>0.000633</td>\n",
       "      <td>Outer</td>\n",
       "      <td>2008</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D156P1_B1</td>\n",
       "      <td>D040M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>-1.519349</td>\n",
       "      <td>0.001699</td>\n",
       "      <td>Outer</td>\n",
       "      <td>1171</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...</td>\n",
       "      <td>D018B3_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FN1</th>\n",
       "      <td>1.396596</td>\n",
       "      <td>0.006217</td>\n",
       "      <td>Outer</td>\n",
       "      <td>1178</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D005P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LRRC17</th>\n",
       "      <td>2.039854</td>\n",
       "      <td>0.011970</td>\n",
       "      <td>Outer</td>\n",
       "      <td>1765</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D007P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNN2</th>\n",
       "      <td>-1.120957</td>\n",
       "      <td>0.013178</td>\n",
       "      <td>Outer</td>\n",
       "      <td>669</td>\n",
       "      <td>62</td>\n",
       "      <td>15</td>\n",
       "      <td>D092P1_B1, D001P1_B1, D005P1_B1, D006P1_B1, D0...</td>\n",
       "      <td>D013M1_B1, D018B2_B1, D018B3_B1, D018M1_B1, D0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R12A</th>\n",
       "      <td>-1.870127</td>\n",
       "      <td>0.029501</td>\n",
       "      <td>Outer</td>\n",
       "      <td>2529</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D018P1_B1, D108P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTN1</th>\n",
       "      <td>-1.331235</td>\n",
       "      <td>0.032099</td>\n",
       "      <td>Outer</td>\n",
       "      <td>51</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D090P1_B1</td>\n",
       "      <td>D109M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DCN</th>\n",
       "      <td>1.383715</td>\n",
       "      <td>0.037626</td>\n",
       "      <td>Outer</td>\n",
       "      <td>807</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>D059F1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRT10</th>\n",
       "      <td>1.034070</td>\n",
       "      <td>0.040717</td>\n",
       "      <td>Outer</td>\n",
       "      <td>1669</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D014P1_B1, D138P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCERG1L</th>\n",
       "      <td>4.724461</td>\n",
       "      <td>0.044047</td>\n",
       "      <td>Outer</td>\n",
       "      <td>3209</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D105P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TSNAX</th>\n",
       "      <td>2.363483</td>\n",
       "      <td>0.532398</td>\n",
       "      <td>Outer</td>\n",
       "      <td>3430</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D126P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL12A1</th>\n",
       "      <td>1.408665</td>\n",
       "      <td>0.589038</td>\n",
       "      <td>Outer</td>\n",
       "      <td>683</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "      <td>D122P1_B1, D105P1_B1, D088P1_B1, D138P1_B1</td>\n",
       "      <td>D020M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>THBS2</th>\n",
       "      <td>1.188005</td>\n",
       "      <td>0.958180</td>\n",
       "      <td>Outer</td>\n",
       "      <td>3256</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D070P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          avg_log2FC  p_val_adj bladder_region  index  patient_count  \\\n",
       "gene                                                                   \n",
       "MYH11      -5.201319   0.000408          Outer   1997              1   \n",
       "MYLK       -2.759718   0.000437          Outer   2009              1   \n",
       "COL1A2      1.253755   0.000439          Outer    687              1   \n",
       "ACTG2      -3.668688   0.000524          Outer     47              3   \n",
       "MYL9       -2.319767   0.000633          Outer   2008              1   \n",
       "FLNA       -1.519349   0.001699          Outer   1171              5   \n",
       "FN1         1.396596   0.006217          Outer   1178              1   \n",
       "LRRC17      2.039854   0.011970          Outer   1765              1   \n",
       "CNN2       -1.120957   0.013178          Outer    669             62   \n",
       "PPP1R12A   -1.870127   0.029501          Outer   2529              2   \n",
       "ACTN1      -1.331235   0.032099          Outer     51              1   \n",
       "DCN         1.383715   0.037626          Outer    807              0   \n",
       "KRT10       1.034070   0.040717          Outer   1669              2   \n",
       "TCERG1L     4.724461   0.044047          Outer   3209              1   \n",
       "TSNAX       2.363483   0.532398          Outer   3430              1   \n",
       "COL12A1     1.408665   0.589038          Outer    683              4   \n",
       "THBS2       1.188005   0.958180          Outer   3256              1   \n",
       "\n",
       "          family_count                                        patient_ids  \\\n",
       "gene                                                                        \n",
       "MYH11                1                                          D013P1_B1   \n",
       "MYLK                 0                                          D165P1_B1   \n",
       "COL1A2               0                                          D150P1_B1   \n",
       "ACTG2                0                    D034P1_B1, D074P1_B1, D065P1_B1   \n",
       "MYL9                 1                                          D156P1_B1   \n",
       "FLNA                 1  D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...   \n",
       "FN1                  0                                          D005P1_B1   \n",
       "LRRC17               0                                          D007P1_B1   \n",
       "CNN2                15  D092P1_B1, D001P1_B1, D005P1_B1, D006P1_B1, D0...   \n",
       "PPP1R12A             0                               D018P1_B1, D108P1_B1   \n",
       "ACTN1                1                                          D090P1_B1   \n",
       "DCN                  1                                                      \n",
       "KRT10                0                               D014P1_B1, D138P1_B1   \n",
       "TCERG1L              0                                          D105P1_B1   \n",
       "TSNAX                0                                          D126P1_B1   \n",
       "COL12A1              1         D122P1_B1, D105P1_B1, D088P1_B1, D138P1_B1   \n",
       "THBS2                0                                          D070P1_B1   \n",
       "\n",
       "                                                 family_ids  \n",
       "gene                                                         \n",
       "MYH11                                             D013M1_B1  \n",
       "MYLK                                                         \n",
       "COL1A2                                                       \n",
       "ACTG2                                                        \n",
       "MYL9                                              D040M1_B1  \n",
       "FLNA                                              D018B3_B1  \n",
       "FN1                                                          \n",
       "LRRC17                                                       \n",
       "CNN2      D013M1_B1, D018B2_B1, D018B3_B1, D018M1_B1, D0...  \n",
       "PPP1R12A                                                     \n",
       "ACTN1                                             D109M1_B1  \n",
       "DCN                                               D059F1_B1  \n",
       "KRT10                                                        \n",
       "TCERG1L                                                      \n",
       "TSNAX                                                        \n",
       "COL12A1                                           D020M1_B1  \n",
       "THBS2                                                        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(17, 8)\n",
      "Vascular\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "      <th>index</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>family_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [avg_log2FC, p_val_adj, bladder_region, index, patient_count, family_count, patient_ids, family_ids]\n",
       "Index: []"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(0, 8)\n",
      "Whole\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "      <th>index</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>family_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MYH11</th>\n",
       "      <td>-4.883850</td>\n",
       "      <td>1.014217e-16</td>\n",
       "      <td>Whole</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D013P1_B1</td>\n",
       "      <td>D013M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTG2</th>\n",
       "      <td>-3.982674</td>\n",
       "      <td>1.786189e-16</td>\n",
       "      <td>Whole</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>D034P1_B1, D074P1_B1, D065P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL9</th>\n",
       "      <td>-2.544347</td>\n",
       "      <td>5.061845e-13</td>\n",
       "      <td>Whole</td>\n",
       "      <td>2008</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D156P1_B1</td>\n",
       "      <td>D040M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYLK</th>\n",
       "      <td>-2.767663</td>\n",
       "      <td>5.469357e-12</td>\n",
       "      <td>Whole</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D165P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R12A</th>\n",
       "      <td>-1.753351</td>\n",
       "      <td>5.688667e-07</td>\n",
       "      <td>Whole</td>\n",
       "      <td>2529</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D018P1_B1, D108P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>-1.203795</td>\n",
       "      <td>6.005021e-06</td>\n",
       "      <td>Whole</td>\n",
       "      <td>1171</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...</td>\n",
       "      <td>D018B3_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CNN2</th>\n",
       "      <td>-1.017152</td>\n",
       "      <td>6.833920e-05</td>\n",
       "      <td>Whole</td>\n",
       "      <td>669</td>\n",
       "      <td>62</td>\n",
       "      <td>15</td>\n",
       "      <td>D092P1_B1, D001P1_B1, D005P1_B1, D006P1_B1, D0...</td>\n",
       "      <td>D013M1_B1, D018B2_B1, D018B3_B1, D018M1_B1, D0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LMOD1</th>\n",
       "      <td>-2.932242</td>\n",
       "      <td>2.717616e-03</td>\n",
       "      <td>Whole</td>\n",
       "      <td>1740</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D098P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SLC38A1</th>\n",
       "      <td>-3.335785</td>\n",
       "      <td>3.194307e-03</td>\n",
       "      <td>Whole</td>\n",
       "      <td>2964</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>D109P1_B1</td>\n",
       "      <td>D109M1_B1, D109M3_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KRT6A</th>\n",
       "      <td>2.529611</td>\n",
       "      <td>3.805985e-03</td>\n",
       "      <td>Whole</td>\n",
       "      <td>1678</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D116P1_B1, D119P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R12B</th>\n",
       "      <td>-1.175030</td>\n",
       "      <td>4.889440e-03</td>\n",
       "      <td>Whole</td>\n",
       "      <td>2530</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D137P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AOC3</th>\n",
       "      <td>-2.056904</td>\n",
       "      <td>8.730573e-03</td>\n",
       "      <td>Whole</td>\n",
       "      <td>196</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D071P1_B1, D102P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          avg_log2FC     p_val_adj bladder_region  index  patient_count  \\\n",
       "gene                                                                      \n",
       "MYH11      -4.883850  1.014217e-16          Whole   1997              1   \n",
       "ACTG2      -3.982674  1.786189e-16          Whole     47              3   \n",
       "MYL9       -2.544347  5.061845e-13          Whole   2008              1   \n",
       "MYLK       -2.767663  5.469357e-12          Whole   2009              1   \n",
       "PPP1R12A   -1.753351  5.688667e-07          Whole   2529              2   \n",
       "FLNA       -1.203795  6.005021e-06          Whole   1171              5   \n",
       "CNN2       -1.017152  6.833920e-05          Whole    669             62   \n",
       "LMOD1      -2.932242  2.717616e-03          Whole   1740              1   \n",
       "SLC38A1    -3.335785  3.194307e-03          Whole   2964              1   \n",
       "KRT6A       2.529611  3.805985e-03          Whole   1678              2   \n",
       "PPP1R12B   -1.175030  4.889440e-03          Whole   2530              1   \n",
       "AOC3       -2.056904  8.730573e-03          Whole    196              2   \n",
       "\n",
       "          family_count                                        patient_ids  \\\n",
       "gene                                                                        \n",
       "MYH11                1                                          D013P1_B1   \n",
       "ACTG2                0                    D034P1_B1, D074P1_B1, D065P1_B1   \n",
       "MYL9                 1                                          D156P1_B1   \n",
       "MYLK                 0                                          D165P1_B1   \n",
       "PPP1R12A             0                               D018P1_B1, D108P1_B1   \n",
       "FLNA                 1  D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...   \n",
       "CNN2                15  D092P1_B1, D001P1_B1, D005P1_B1, D006P1_B1, D0...   \n",
       "LMOD1                0                                          D098P1_B1   \n",
       "SLC38A1              2                                          D109P1_B1   \n",
       "KRT6A                0                               D116P1_B1, D119P1_B1   \n",
       "PPP1R12B             0                                          D137P1_B1   \n",
       "AOC3                 0                               D071P1_B1, D102P1_B1   \n",
       "\n",
       "                                                 family_ids  \n",
       "gene                                                         \n",
       "MYH11                                             D013M1_B1  \n",
       "ACTG2                                                        \n",
       "MYL9                                              D040M1_B1  \n",
       "MYLK                                                         \n",
       "PPP1R12A                                                     \n",
       "FLNA                                              D018B3_B1  \n",
       "CNN2      D013M1_B1, D018B2_B1, D018B3_B1, D018M1_B1, D0...  \n",
       "LMOD1                                                        \n",
       "SLC38A1                                D109M1_B1, D109M3_B1  \n",
       "KRT6A                                                        \n",
       "PPP1R12B                                                     \n",
       "AOC3                                                         "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(12, 8)\n",
      "subSMC\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>avg_log2FC</th>\n",
       "      <th>p_val_adj</th>\n",
       "      <th>bladder_region</th>\n",
       "      <th>index</th>\n",
       "      <th>patient_count</th>\n",
       "      <th>family_count</th>\n",
       "      <th>patient_ids</th>\n",
       "      <th>family_ids</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>KRT10</th>\n",
       "      <td>-1.121357</td>\n",
       "      <td>0.000085</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>1669</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D014P1_B1, D138P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYH11</th>\n",
       "      <td>1.256399</td>\n",
       "      <td>0.001776</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>1997</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D013P1_B1</td>\n",
       "      <td>D013M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A2</th>\n",
       "      <td>-0.870309</td>\n",
       "      <td>0.005751</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>687</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D150P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYLK</th>\n",
       "      <td>0.991432</td>\n",
       "      <td>0.011603</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>2009</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D165P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACTG2</th>\n",
       "      <td>1.173604</td>\n",
       "      <td>0.049401</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>47</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>D034P1_B1, D074P1_B1, D065P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FLNA</th>\n",
       "      <td>0.753236</td>\n",
       "      <td>0.092131</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>1171</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...</td>\n",
       "      <td>D018B3_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYL9</th>\n",
       "      <td>0.962548</td>\n",
       "      <td>0.097923</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>2008</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>D156P1_B1</td>\n",
       "      <td>D040M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SPTBN1</th>\n",
       "      <td>-0.848132</td>\n",
       "      <td>0.984124</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>3093</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td></td>\n",
       "      <td>D179M1_B1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PPP1R12A</th>\n",
       "      <td>0.911623</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>2529</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>D018P1_B1, D108P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FNDC3A</th>\n",
       "      <td>-2.164905</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>subSMC</td>\n",
       "      <td>1181</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>D127P1_B1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          avg_log2FC  p_val_adj bladder_region  index  patient_count  \\\n",
       "gene                                                                   \n",
       "KRT10      -1.121357   0.000085         subSMC   1669              2   \n",
       "MYH11       1.256399   0.001776         subSMC   1997              1   \n",
       "COL1A2     -0.870309   0.005751         subSMC    687              1   \n",
       "MYLK        0.991432   0.011603         subSMC   2009              1   \n",
       "ACTG2       1.173604   0.049401         subSMC     47              3   \n",
       "FLNA        0.753236   0.092131         subSMC   1171              5   \n",
       "MYL9        0.962548   0.097923         subSMC   2008              1   \n",
       "SPTBN1     -0.848132   0.984124         subSMC   3093              0   \n",
       "PPP1R12A    0.911623   1.000000         subSMC   2529              2   \n",
       "FNDC3A     -2.164905   1.000000         subSMC   1181              1   \n",
       "\n",
       "          family_count                                        patient_ids  \\\n",
       "gene                                                                        \n",
       "KRT10                0                               D014P1_B1, D138P1_B1   \n",
       "MYH11                1                                          D013P1_B1   \n",
       "COL1A2               0                                          D150P1_B1   \n",
       "MYLK                 0                                          D165P1_B1   \n",
       "ACTG2                0                    D034P1_B1, D074P1_B1, D065P1_B1   \n",
       "FLNA                 1  D081P1_B1, D088P1_B1, D019P1_B1, D019P2_B1, D0...   \n",
       "MYL9                 1                                          D156P1_B1   \n",
       "SPTBN1               1                                                      \n",
       "PPP1R12A             0                               D018P1_B1, D108P1_B1   \n",
       "FNDC3A               0                                          D127P1_B1   \n",
       "\n",
       "         family_ids  \n",
       "gene                 \n",
       "KRT10                \n",
       "MYH11     D013M1_B1  \n",
       "COL1A2               \n",
       "MYLK                 \n",
       "ACTG2                \n",
       "FLNA      D018B3_B1  \n",
       "MYL9      D040M1_B1  \n",
       "SPTBN1    D179M1_B1  \n",
       "PPP1R12A             \n",
       "FNDC3A               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10, 8)\n"
     ]
    }
   ],
   "source": [
    "#sample_counts_df = filtered_wes_data.groupby(['Gene.refGeneWithVer'], as_index=False)[['patient_count', 'family_count']].sum()\n",
    "sample_counts_df = filtered_wes_data.groupby(['Gene.refGeneWithVer'], as_index=False).agg({\n",
    "    'patient_count': 'sum',\n",
    "    'family_count': 'sum',\n",
    "    'patient_ids': lambda x: ', '.join(\n",
    "        [str(s).strip() for s in x if pd.notna(s) and str(s).strip() != '']),\n",
    "    'family_ids': lambda x: ', '.join(\n",
    "        [str(s).strip() for s in x if pd.notna(s) and str(s).strip() != ''])\n",
    "}).reset_index()\n",
    "\n",
    "sample_counts_df.rename(columns={\"Gene.refGeneWithVer\": \"gene\"}, errors=\"raise\", inplace=True)\n",
    "for groupname, group_df in sliced.groupby(\"bladder_region\"):\n",
    "    print(groupname)\n",
    "    common_genes = pd.concat([group_df.set_index('gene'),sample_counts_df.set_index('gene')], axis=1, join='inner')\n",
    "    display(common_genes)\n",
    "    common_genes.to_csv(f\"/Users/nxr042/research/visium/megabladder/results/all_int_bl_no_mgb1/WES-qiushi/{groupname}-WES-{af_cut}_patient_counts_df.csv\")\n",
    "    print(common_genes.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "5631b5ec-3f71-4bbc-9c47-4e63a2606585",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Common genes: {'LMOD1', 'IGFBP2', 'KRT6A', 'ACTG2', 'POLR1A', 'FN1', 'DCN', 'TCERG1L', 'PPP1R12A', 'THBS2', 'MYLK', 'LRRC17', 'ARHGEF11', 'TNNT2', 'AOC3', 'FNDC3A', 'ACTN1', 'SLC38A1', 'TSNAX', 'KRT1', 'CNN2', 'FLNA', 'JPH2', 'COL12A1', 'COL1A2', 'MYL9', 'KRT10', 'CNN1', 'DZIP1L', 'MYH11', 'TMED2', 'SPTBN1', 'PPP1R12B'}\n"
     ]
    }
   ],
   "source": [
    "unique_genes_df1 = sliced['gene'].dropna().unique()\n",
    "unique_genes_df2 = sample_counts_df['gene'].dropna().unique()\n",
    "\n",
    "# Find intersection (genes present in both)\n",
    "common_genes = set(unique_genes_df1) & set(unique_genes_df2)\n",
    "print(f\"Common genes: {common_genes}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd5ec0fb-ba17-4d6e-852c-cd5ec15977f9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2cbf5780-ea2f-47a3-bb4c-032697c6807f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59de3836-3b64-4f8c-b6bb-39a6304c60bc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e67785d2-4822-4b84-b77a-2f96fdd4f63c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "nix_env",
   "language": "python",
   "name": "nix_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
